### DISCUSSION TITLE
NetRCInfo.netrc_info() doesn't work when run in a systemd service with DynamicUser=true

### DISCUSSION THREAD

**Discussion by djmattyg007** [encode/httpx#2526] [Potential Issue] [answered]
The function in question, `netrc_info()`, iterates over three paths. Two of these paths refer to the home directory of the current process' user, by prefixing the paths with `~/`. These paths are run through `pathlib.Path.expanduser()`, which raises an exception if the home directory can't be determined.

When the code is running in a systemd service with `DynamicUser=true`, there is no home directory. This means `expanduser()` will always raise an exception.

Fortunately, there's a potential override: set the `NETRC` environment variable to something. As long as it points to a file that exists, and doesn't need to know the current user's home directory, this is an acceptable workaround.

Unfortunately, there's an `is_file()` check performed on the `Path` object before attempting to use it. This means I can't just set `NETRC=/dev/null` as a simple way of working around this problem, even though `netrc.netrc()` will happily accept `/dev/null`.

Why does any of this matter? Because I'm not using `httpx` directly. I'm using the library `python-telegram-bot`, and have no control over how it uses `httpx` internally.

To summarise, there are two problems:

- The call to `pathlib.Path.expanduser()` can raise an exception that isn't currently being caught
- Setting `NETRC=/dev/null` doesn't work as expected, despite it being a perfectly suitable file to use in this context

I'm happy to raise an issue for this, and also to implement a PR to resolve this.

**Comment by lovelydinosaur** [comment, marked as answer]
Thanks for the comprehensive report.

I've escalated this into an issue - https://github.com/encode/httpx/issues/2532
